### PR TITLE
Add setDiag support for any square domain

### PR DESCRIPTION
**Pull request by wongalvis14**
setDiag now supports any regular square domain, as long as it is in the form of {a..#n, b..#n}

### CODE REVIEW

**Comment by wongalvis14** [comment]
@e-kayrakli The old PR is here https://github.com/chapel-lang/chapel/pull/13916 

I've ran the tests myself and verified that the output is correct, but start_test is doing something weird after firing some warnings about locale settings, I'm still looking to fix it but I thought it doesn't have to block this change

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -444,16 +444,19 @@ proc eye(Dom: domain(2), type eltType=real) {
 proc setDiag (ref X: [?D] ?eltType, in k: int = 0, val: eltType = 0)
               where isDenseMatrix(X) {
   var start, end = 0;
+  var (low_i, low_j) = D.low;
+  var (high_i, high_j) = D.high;
   if (k >= 0) { // upper or main diagonal
-    start = 1;
-    end = D.shape(1) - k;
+    start = low_i;
+    end = high_i - k;
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl]: Let's change all `low_i`-like variable names to camelCase. e.g. `lowI`, `highI`... Personally, that looks ugly to me and I'd prefer `iLow`, `iHigh` instead.

**Comment by e-kayrakli** [comment]
Besides my inline comment about variable names, I think this looks good!

Testing status:
- [x] relevant test passes standard local
- [x] relevant test passes standard gasnet
- [x] relevant test passes on XC in performance testing config
- [ ] full standard local, just in case

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -444,16 +444,19 @@ proc eye(Dom: domain(2), type eltType=real) {
 proc setDiag (ref X: [?D] ?eltType, in k: int = 0, val: eltType = 0)
```
Comment (ben-albrecht) [inline modules/packages/LinearAlgebra.chpl]: I think we should check that:

1. `X.rank == 2`
2. `isSquare(X)`

and give a user-friendly error message if not (see other functions for examples of this).

**Comment by ben-albrecht** [comment]
@wongalvis14 - Could you update the PR message better reflect that this is generalizing `setDiag` to support domains with any offset, rather than 1-based domains? Thanks!

**Pull request closed without merging**